# **Waze Project**
**Course 2 - Get Started with Python**

Welcome to the Waze Project!

Your Waze data analytics team is still in the early stages of their user churn project. Previously, you were asked to complete a project proposal by your supervisor, May Santner. You have received notice that your project proposal has been approved and that your team has been given access to Waze's user data. To get clear insights, the user data must be inspected and prepared for the upcoming process of exploratory data analysis (EDA).

A Python notebook has been prepared to guide you through this project. Answer the questions and create an executive summary for the Waze data team.

# **Course 2 End-of-course project: Inspect and analyze data**

In this activity, you will examine data provided and prepare it for analysis. This activity will help ensure the information is,

1.   Ready to answer questions and yield insights

2.   Ready for visualizations

3.   Ready for future hypothesis testing and statistical methods
<br/>

**The purpose** of this project is to investigate and understand the data provided.

**The goal** is to use a dataframe contructed within Python, perform a cursory inspection of the provided dataset, and inform team members of your findings.
<br/>

*This activity has three parts:*

**Part 1:** Understand the situation
* How can you best prepare to understand and organize the provided information?

**Part 2:** Understand the data

* Create a pandas dataframe for data learning, future exploratory data analysis (EDA), and statistical activities

* Compile summary information about the data to inform next steps

**Part 3:** Understand the variables

* Use insights from your examination of the summary data to guide deeper investigation into variables


<br/>

Follow the instructions and answer the following questions to complete the activity. Then, you will complete an Executive Summary using the questions listed on the PACE Strategy Document.

Be sure to complete this activity before moving on. The next course item will provide you with a completed exemplar to compare to your own work.



# **Identify data types and compile summary information**


<img src="images/Pace.png" width="100" height="100" align=left>

# **PACE stages**

Throughout these project notebooks, you'll see references to the problem-solving framework, PACE. The following notebook components are labeled with the respective PACE stages: Plan, Analyze, Construct, and Execute.

<img src="images/Plan.png" width="100" height="100" align=left>


## **PACE: Plan**

Consider the questions in your PACE Strategy Document and those below to craft your response:

### **Task 1. Understand the situation**

*   How can you best prepare to understand and organize the provided driver data?


*Begin by exploring your dataset and consider reviewing the Data Dictionary.*

==> ENTER YOUR RESPONSE HERE

<img src="images/Analyze.png" width="100" height="100" align=left>

## **PACE: Analyze**

Consider the questions in your PACE Strategy Document to reflect on the Analyze stage.

### **Task 2a. Imports and data loading**

Start by importing the packages that you will need to load and explore the dataset. Make sure to use the following import statements:

*   `import pandas as pd`

*   `import numpy as np`


In [1]:
# Import packages for data manipulation
### YOUR CODE HERE ###
import pandas as pd
import numpy as np

Then, load the dataset into a dataframe. Creating a dataframe will help you conduct data manipulation, exploratory data analysis (EDA), and statistical activities.

**Note:** As shown in this cell, the dataset has been automatically loaded in for you. You do not need to download the .csv file, or provide more code, in order to access the dataset and proceed with this lab. Please continue with this activity by completing the following instructions.

In [2]:
# Load dataset into dataframe
df = pd.read_csv('waze_dataset.csv')

### **Task 2b. Summary information**

View and inspect summary information about the dataframe by **coding the following:**

1.   df.head(10)
2.   df.info()

*Consider the following questions:*

1. When reviewing the `df.head()` output, are there any variables that have missing values?

2. When reviewing the `df.info()` output, what are the data types? How many rows and columns do you have?

3. Does the dataset have any missing values?

In [3]:
### YOUR CODE HERE ###
df.head(10)

ID     label  sessions  drives  total_sessions  n_days_after_onboarding  \
0   0  retained       283     226      296.748273                     2276   
1   1  retained       133     107      326.896596                     1225   
2   2  retained       114      95      135.522926                     2651   
3   3  retained        49      40       67.589221                       15   
4   4  retained        84      68      168.247020                     1562   
5   5  retained       113     103      279.544437                     2637   
6   6  retained         3       2      236.725314                      360   
7   7  retained        39      35      176.072845                     2999   
8   8  retained        57      46      183.532018                      424   
9   9   churned        84      68      244.802115                     2997   

   total_navigations_fav1  total_navigations_fav2  driven_km_drives  \
0                     208                       0       2628.845068   
1                      19                      64      13715.920550   
2                       0                       0       3059.148818   
3                     322                       7        913.591123   
4                     166                       5       3950.202008   
5                       0                       0        901.238699   
6                     185                      18       5249.172828   
7                       0                       0       7892.052468   
8                       0                      26       2651.709764   
9                      72                       0       6043.460295   

   duration_minutes_drives  activity_days  driving_days   device  
0              1985.775061             28            19  Android  
1              3160.472914             13            11   iPhone  
2              1610.735904             14             8  Android  
3               587.196542              7             3   iPhone  
4              1219.555924             27            18  Android  
5               439.101397             15            11   iPhone  
6               726.577205             28            23   iPhone  
7              2466.981741             22            20   iPhone  
8              1594.342984             25            20  Android  
9              2341.838528              7             3   iPhone

In [4]:
### YOUR CODE HERE ###
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14999 entries, 0 to 14998
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   ID                       14999 non-null  int64  
 1   label                    14299 non-null  object 
 2   sessions                 14999 non-null  int64  
 3   drives                   14999 non-null  int64  
 4   total_sessions           14999 non-null  float64
 5   n_days_after_onboarding  14999 non-null  int64  
 6   total_navigations_fav1   14999 non-null  int64  
 7   total_navigations_fav2   14999 non-null  int64  
 8   driven_km_drives         14999 non-null  float64
 9   duration_minutes_drives  14999 non-null  float64
 10  activity_days            14999 non-null  int64  
 11  driving_days             14999 non-null  int64  
 12  device                   14999 non-null  object 
dtypes: float64(3), int64(8), object(2)
memory usage: 1.5+ MB


==> ENTER YOUR RESPONSES TO QUESTIONS 1-3 HERE

### **Task 2c. Null values and summary statistics**

Compare the summary statistics of the 700 rows that are missing labels with summary statistics of the rows that are not missing any values.

**Question:** Is there a discernible difference between the two populations?


In [5]:
# Isolate rows with null values
### YOUR CODE HERE ###
df.describe()
# Display summary stats of rows with null values
### YOUR CODE HERE ###


ID      sessions        drives  total_sessions  \
count  14999.000000  14999.000000  14999.000000    14999.000000   
mean    7499.000000     80.633776     67.281152      189.964447   
std     4329.982679     80.699065     65.913872      136.405128   
min        0.000000      0.000000      0.000000        0.220211   
25%     3749.500000     23.000000     20.000000       90.661156   
50%     7499.000000     56.000000     48.000000      159.568115   
75%    11248.500000    112.000000     93.000000      254.192341   
max    14998.000000    743.000000    596.000000     1216.154633   

       n_days_after_onboarding  total_navigations_fav1  \
count             14999.000000            14999.000000   
mean               1749.837789              121.605974   
std                1008.513876              148.121544   
min                   4.000000                0.000000   
25%                 878.000000                9.000000   
50%                1741.000000               71.000000   
75%                2623.500000              178.000000   
max                3500.000000             1236.000000   

       total_navigations_fav2  driven_km_drives  duration_minutes_drives  \
count            14999.000000      14999.000000             14999.000000   
mean                29.672512       4039.340921              1860.976012   
std                 45.394651       2502.149334              1446.702288   
min                  0.000000         60.441250                18.282082   
25%                  0.000000       2212.600607               835.996260   
50%                  9.000000       3493.858085              1478.249859   
75%                 43.000000       5289.861262              2464.362632   
max                415.000000      21183.401890             15851.727160   

       activity_days  driving_days  
count   14999.000000  14999.000000  
mean       15.537102     12.179879  
std         9.004655      7.824036  
min         0.000000      0.000000  
25%         8.000000      5.000000  
50%        16.000000     12.000000  
75%        23.000000     19.000000  
max        31.000000     30.000000

In [9]:
# Isolate rows without null values
### YOUR CODE HERE ###
df_no_nulls = df.dropna()
# Display summary stats of rows without null values
### YOUR CODE HERE ###
df_no_nulls.describe()

ID      sessions        drives  total_sessions  \
count  14299.000000  14299.000000  14299.000000    14299.000000   
mean    7503.573117     80.623820     67.255822      189.547409   
std     4331.207621     80.736502     65.947295      136.189764   
min        0.000000      0.000000      0.000000        0.220211   
25%     3749.500000     23.000000     20.000000       90.457733   
50%     7504.000000     56.000000     48.000000      158.718571   
75%    11257.500000    111.000000     93.000000      253.540450   
max    14998.000000    743.000000    596.000000     1216.154633   

       n_days_after_onboarding  total_navigations_fav1  \
count             14299.000000            14299.000000   
mean               1751.822505              121.747395   
std                1008.663834              147.713428   
min                   4.000000                0.000000   
25%                 878.500000               10.000000   
50%                1749.000000               71.000000   
75%                2627.500000              178.000000   
max                3500.000000             1236.000000   

       total_navigations_fav2  driven_km_drives  duration_minutes_drives  \
count            14299.000000      14299.000000             14299.000000   
mean                29.638296       4044.401535              1864.199794   
std                 45.350890       2504.977970              1448.005047   
min                  0.000000         60.441250                18.282082   
25%                  0.000000       2217.319909               840.181344   
50%                  9.000000       3496.545617              1479.394387   
75%                 43.000000       5299.972162              2466.928876   
max                415.000000      21183.401890             15851.727160   

       activity_days  driving_days  
count   14299.000000  14299.000000  
mean       15.544653     12.182530  
std         9.016088      7.833835  
min         0.000000      0.000000  
25%         8.000000      5.000000  
50%        16.000000     12.000000  
75%        23.000000     19.000000  
max        31.000000     30.000000

==> ENTER YOUR RESPONSE HERE

### **Task 2d. Null values - device counts**

Next, check the two populations with respect to the `device` variable.

**Question:** How many iPhone users had null values and how many Android users had null values?

In [11]:
# Get count of null values by device
### YOUR CODE HERE ###
null_counts_by_device = df.groupby('device').apply(lambda x: x.isnull().sum()) #count of null values
null_counts_by_device

ID  label  sessions  drives  total_sessions  n_days_after_onboarding  \
device                                                                          
Android   0    253         0       0               0                        0   
iPhone    0    447         0       0               0                        0   

         total_navigations_fav1  total_navigations_fav2  driven_km_drives  \
device                                                                      
Android                       0                       0                 0   
iPhone                        0                       0                 0   

         duration_minutes_drives  activity_days  driving_days  device  
device                                                                 
Android                        0              0             0       0  
iPhone                         0              0             0       0

==> ENTER YOUR RESPONSE HERE

Now, of the rows with null values, calculate the percentage with each device&mdash;Android and iPhone. You can do this directly with the [`value_counts()`](https://pandas.pydata.org/docs/reference/api/pandas.Series.value_counts.html) function.

In [ ]:
# Calculate % of iPhone nulls and Android nulls
### YOUR CODE HERE ###
df_with_label = df
df_with_label 

How does this compare to the device ratio in the full dataset?

In [15]:
# Calculate % of iPhone users and Android users in full dataset
### YOUR CODE HERE ###
device_counts = df['device'].value_counts()
device_percentages = (device_counts / len(df)) * 100
device_percentages

iPhone     64.484299
Android    35.515701
Name: device, dtype: float64

The percentage of missing values by each device is consistent with their representation in the data overall.

There is nothing to suggest a non-random cause of the missing data.

Examine the counts and percentages of users who churned vs. those who were retained. How many of each group are represented in the data?

In [19]:
# Calculate counts of churned vs. retained
### YOUR CODE HERE ###
df_without_null = df.dropna()
summary_df = df_without_null.groupby('label').agg(
    count=('ID', 'size'),
    percentage=('ID', lambda x: len(x) / len(df_without_null) * 100)
).reset_index()

summary_df

label  count  percentage
0   churned   2536   17.735506
1  retained  11763   82.264494

This dataset contains 82% retained users and 18% churned users.

Next, compare the medians of each variable for churned and retained users. The reason for calculating the median and not the mean is that you don't want outliers to unduly affect the portrayal of a typical user. Notice, for example, that the maximum value in the `driven_km_drives` column is 21,183 km. That's more than half the circumference of the earth!

In [21]:
# Calculate median values of all columns for churned and retained users
### YOUR CODE HERE ###
grouped_by_label = df_without_null.groupby(['label']).agg([ 'median'])
grouped_by_label

ID sessions drives total_sessions n_days_after_onboarding  \
          median   median median         median                  median   
label                                                                     
churned   7477.5     59.0   50.0     164.339042                  1321.0   
retained  7509.0     56.0   47.0     157.586756                  1843.0   

         total_navigations_fav1 total_navigations_fav2 driven_km_drives  \
                         median                 median           median   
label                                                                     
churned                    84.5                   11.0      3652.655666   
retained                   68.0                    9.0      3464.684614   

         duration_minutes_drives activity_days driving_days  
                          median        median       median  
label                                                        
churned              1607.183785           8.0          6.0  
retained             1458.046141          17.0         14.0

This offers an interesting snapshot of the two groups, churned vs. retained:

Users who churned averaged ~3 more drives in the last month than retained users, but retained users used the app on over twice as many days as churned users in the same time period.

The median churned user drove ~200 more kilometers and 2.5 more hours during the last month than the median retained user.

It seems that churned users had more drives in fewer days, and their trips were farther and longer in duration. Perhaps this is suggestive of a user profile. Continue exploring!

Calculate the median kilometers per drive in the last month for both retained and churned users.

Begin by dividing the `driven_km_drives` column by the `drives` column. Then, group the results by churned/retained and calculate the median km/drive of each group.

In [24]:
df_without_null.head()

ID     label  sessions  drives  total_sessions  n_days_after_onboarding  \
0   0  retained       283     226      296.748273                     2276   
1   1  retained       133     107      326.896596                     1225   
2   2  retained       114      95      135.522926                     2651   
3   3  retained        49      40       67.589221                       15   
4   4  retained        84      68      168.247020                     1562   

   total_navigations_fav1  total_navigations_fav2  driven_km_drives  \
0                     208                       0       2628.845068   
1                      19                      64      13715.920550   
2                       0                       0       3059.148818   
3                     322                       7        913.591123   
4                     166                       5       3950.202008   

   duration_minutes_drives  activity_days  driving_days   device  \
0              1985.775061             28            19  Android   
1              3160.472914             13            11   iPhone   
2              1610.735904             14             8  Android   
3               587.196542              7             3   iPhone   
4              1219.555924             27            18  Android   

   kilo_per_drive  
0       11.632058  
1      128.186173  
2       32.201567  
3       22.839778  
4       58.091206

In [30]:
# Add a column to df called `km_per_drive`
### YOUR CODE HERE ###
km_per_drive = df['driven_km_drives']  / df['drives'] 
km_per_drive
df['km_per_drive1'] = km_per_drive
# Group by `label`, calculate the median, and isolate for km per drive
### YOUR CODE HERE ###
median_km_per_drive = df.groupby('label')['km_per_drive1'].median()

In [31]:
median_km_per_drive

label
churned     74.109416
retained    75.014702
Name: km_per_drive1, dtype: float64

The median retained user drove about one more kilometer per drive than the median churned user. How many kilometers per driving day was this?

To calculate this statistic, repeat the steps above using `driving_days` instead of `drives`.

In [32]:
# Add a column to df called `km_per_driving_day`
# Add a column to df called `km_per_drive`
### YOUR CODE HERE ###
km_per_drive_days = df['driven_km_drives']  / df['driving_days'] 
km_per_drive_days
df['km_per_driving_days'] = km_per_drive_days
# Group by `label`, calculate the median, and isolate for km per drive
### YOUR CODE HERE ###
median_km_per_drive_days = df.groupby('label')['km_per_driving_days'].median()
median_km_per_drive_days

label
churned     697.541999
retained    289.549333
Name: km_per_driving_days, dtype: float64

Now, calculate the median number of drives per driving day for each group.

In [34]:
# Add a column to df called `drives_per_driving_day`
### YOUR CODE HERE ###
df['drives_per_driving_day'] = df['drives'] / df['driving_days']
# Group by `label`, calculate the median, and isolate for drives per driving day
### YOUR CODE HERE ###
mediam_drives_per_driving_day = df.groupby('label')['drives_per_driving_day'].median()
mediam_drives_per_driving_day

label
churned     10.0000
retained     4.0625
Name: drives_per_driving_day, dtype: float64

The median user who churned drove 698 kilometers each day they drove last month, which is almost ~240% the per-drive-day distance of retained users. The median churned user had a similarly disproporionate number of drives per drive day compared to retained users.

It is clear from these figures that, regardless of whether a user churned or not, the users represented in this data are serious drivers! It would probably be safe to assume that this data does not represent typical drivers at large. Perhaps the data&mdash;and in particular the sample of churned users&mdash;contains a high proportion of long-haul truckers.

In consideration of how much these users drive, it would be worthwhile to recommend to Waze that they gather more data on these super-drivers. It's possible that the reason for their driving so much is also the reason why the Waze app does not meet their specific set of needs, which may differ from the needs of a more typical driver, such as a commuter.

Finally, examine whether there is an imbalance in how many users churned by device type.

Begin by getting the overall counts of each device type for each group, churned and retained.

In [36]:
# For each label, calculate the number of Android users and iPhone users
### YOUR CODE HERE ###
grouped_by_label_device = df.groupby(['label', 'device']).agg(['count','mean', 'median'])
grouped_by_label_device

ID                      sessions                   drives  \
                 count         mean  median    count       mean median  count   
label    device                                                                 
churned  Android   891  7553.647587  7579.0      891  89.601571   59.0    891   
         iPhone   1645  7540.089362  7434.0     1645  85.959271   59.0   1645   
retained Android  4183  7582.540521  7596.0     4183  76.801339   55.0   4183   
         iPhone   7580  7446.184433  7476.0     7580  80.520053   56.0   7580   

                                   total_sessions  ... driving_days  \
                       mean median          count  ...       median   
label    device                                    ...                
churned  Android  74.544332   50.0            891  ...          6.0   
         iPhone   71.748328   50.0           1645  ...          5.0   
retained Android  64.209419   46.0           4183  ...         14.0   
         iPhone   67.105277   48.0           7580  ...         14.0   

                 km_per_drive1                 km_per_driving_days       \
                         count mean     median               count mean   
label    device                                                           
churned  Android           891  NaN  74.981626                 891  NaN   
         iPhone           1645  NaN  73.377717                1645  NaN   
retained Android          4183  NaN  76.625683                4183  NaN   
         iPhone           7580  NaN  74.187745                7580  NaN   

                             drives_per_driving_day                  
                      median                  count mean     median  
label    device                                                      
churned  Android  657.469602                    890  NaN   9.166667  
         iPhone   715.357995                   1644  NaN  10.387500  
retained Android  288.761975                   4180  NaN   4.000000  
         iPhone   290.198789                   7578  NaN   4.133333  

[4 rows x 42 columns]

Now, within each group, churned and retained, calculate what percent was Android and what percent was iPhone.

In [38]:
# For each label, calculate the percentage of Android users and iPhone users
### YOUR CODE HERE ###
df_without_null = df.dropna()

# Step 1: Group by label and user_type
grouped = df.groupby(['label', 'device']).size().reset_index(name='count')

# Step 2: Pivot the data
pivot = grouped.pivot(index='label', columns='device', values='count').fillna(0)

# Step 3: Calculate percentages
pivot['total'] = pivot.sum(axis=1)
pivot['Android_percent'] = (pivot['Android'] / pivot['total']) * 100
pivot['iPhone_percent'] = (pivot['iPhone'] / pivot['total']) * 100

# Drop the total column if it's not needed
pivot = pivot.drop(columns='total')

# Print the result
print(pivot)

device    Android  iPhone  Android_percent  iPhone_percent
label                                                     
churned       891    1645        35.134069       64.865931
retained     4183    7580        35.560656       64.439344


The ratio of iPhone users and Android users is consistent between the churned group and the retained group, and those ratios are both consistent with the ratio found in the overall dataset.

<img src="images/Construct.png" width="100" height="100" align=left>

## **PACE: Construct**

**Note**: The Construct stage does not apply to this workflow. The PACE framework can be adapted to fit the specific requirements of any project.



<img src="images/Execute.png" width="100" height="100" align=left>

## **PACE: Execute**

Consider the questions in your PACE Strategy Document and those below to craft your response:

### **Task 3. Conclusion**

Recall that your supervisor, May Santer, asked you to share your findings with the data team in an executive summary. Consider the following questions as you prepare to write your summary. Think about key points you may want to share with the team, and what information is most relevant to the user churn project.

**Questions:**

1. Did the data contain any missing values? How many, and which variables were affected? Was there a pattern to the missing data?

2. What is a benefit of using the median value of a sample instead of the mean?

3. Did your investigation give rise to further questions that you would like to explore or ask the Waze team about?

4. What percentage of the users in the dataset were Android users and what percentage were iPhone users?

5. What were some distinguishing characteristics of users who churned vs. users who were retained?

6. Was there an appreciable difference in churn rate between iPhone users vs. Android users?





**Congratulations!** You've completed this lab. However, you may not notice a green check mark next to this item on Coursera's platform. Please continue your progress regardless of the check mark. Just click on the "save" icon at the top of this notebook to ensure your work has been logged.